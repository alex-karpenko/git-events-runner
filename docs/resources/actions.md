# Actions

TBW

## Action

TBW

## ClusterAction

TBW
