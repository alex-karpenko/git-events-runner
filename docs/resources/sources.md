# Sources

TBW

## GitRepo

TBW

## ClusterGitRepo

TBW
