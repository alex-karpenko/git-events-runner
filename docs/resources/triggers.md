# Triggers

Both types of the triggers are almost identical except one principal section:

* `ScheduleTrigger` has a **schedule** section which describes how often it should check sources for changes.
* `WebhookTrigger` has a **webhook** section which describes triggers behavior on HTTP requests.

**sources** section contains list (one or more) of sources to check.
Every time ScheduleTrigger runs, it checks all sources one by one and creates Job for each source that has been changed
from the previous check.

It's possible to specify additional constrain on the source check: list of the file glob patterns, that should be
changed to get the trigger fired.
Adding or removing files, as well as changing of any file content (hash), means that source has been changed.
Pattern format is a well-known `.gitignore` format, described in details in
the [git documentation](https://git-scm.com/docs/gitignore#_pattern_format).

WebhookTrigger can be called by specifying path to the trigger in one of the two forms:

* By trigger name without a source name: `/namespace/trigger`, like `/default/trigger-example`.
  In this case, trigger checks all defined sources one by one like a ScheduleTrigger.
* With full source name: /namespace/trigger/source, like `/default/trigger-example/source-repo-1`.
  In this case, trigger check specified source only.

Second form is disabled by default to avoid possible unpredictable workload burst, and may be enabled by `multiSource`
flag.

If `authConfig` section of the WebhookTrigger is defined,
authorization header should be present in each request to the trigger.
Header name may be changed.

## ScheduleTrigger

```yaml title="ScheduleTrigger"
--8<-- "docs/examples/scheduletrigger.yaml"
```

## WebhookTrigger

```yaml title="WebhookTrigger"
--8<-- "docs/examples/webhooktrigger.yaml"
```
