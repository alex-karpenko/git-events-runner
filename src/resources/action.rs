use super::{
    random_string,
    trigger::{TriggerGitRepoReference, TriggerSourceKind},
    CustomApiResource,
};
use crate::{config::RuntimeConfig, Error, Result};
use chrono::{DateTime, Local};
use k8s_openapi::{
    api::{
        batch::v1::{Job, JobSpec},
        core::v1::{
            Affinity, Container, EmptyDirVolumeSource, EnvVar, PodSpec, PodTemplateSpec, Toleration, Volume,
            VolumeMount,
        },
    },
    apimachinery::pkg::apis::meta::v1::OwnerReference,
};
use kube::{
    api::{ObjectMeta, PostParams},
    Api, Client, CustomResource, Resource, ResourceExt,
};
use schemars::JsonSchema;
use serde::{Deserialize, Serialize};
use std::{collections::BTreeMap, time::SystemTime};
use strum::{Display, EnumString};
use tracing::info;

pub const ACTION_JOB_IDENTITY_LABEL: &str = "git-events-runner.rs/controller-identity";
pub const ACTION_JOB_ACTION_KIND_LABEL: &str = "git-events-runner.rs/action-kind";
pub const ACTION_JOB_ACTION_NAME_LABEL: &str = "git-events-runner.rs/action-name";
pub const ACTION_JOB_SOURCE_KIND_LABEL: &str = "git-events-runner.rs/source-kind";
pub const ACTION_JOB_SOURCE_NAME_LABEL: &str = "git-events-runner.rs/source-name";

#[derive(CustomResource, Deserialize, Serialize, Clone, Debug, JsonSchema, PartialEq)]
#[cfg_attr(test, derive(Default))]
#[kube(
    kind = "Action",
    group = "git-events-runner.rs",
    version = "v1alpha1",
    doc = "Action custom resource definition",
    namespaced
)]
#[serde(rename_all = "camelCase")]
pub struct ActionSpec {
    #[serde(skip_serializing_if = "Option::is_none")]
    source_override: Option<ActionSourceOverride>,
    #[serde(default)]
    action_job: ActionJob,
}

#[derive(CustomResource, Deserialize, Serialize, Clone, Debug, JsonSchema, PartialEq)]
#[cfg_attr(test, derive(Default))]
#[kube(
    kind = "ClusterAction",
    group = "git-events-runner.rs",
    version = "v1alpha1",
    doc = "ClusterAction custom resource definition"
)]
#[serde(rename_all = "camelCase")]
pub struct ClusterActionSpec {
    #[serde(skip_serializing_if = "Option::is_none")]
    source_override: Option<ClusterActionSourceOverride>,
    #[serde(default)]
    action_job: ActionJob,
}

#[derive(Deserialize, Serialize, Clone, Default, Debug, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ActionSourceOverride {
    kind: TriggerSourceKind,
    name: String,
    reference: TriggerGitRepoReference,
}

#[derive(Deserialize, Serialize, Clone, Default, Debug, JsonSchema, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct ClusterActionSourceOverride {
    kind: TriggerClusterSourceKind,
    name: String,
    reference: TriggerGitRepoReference,
}

#[derive(Deserialize, Serialize, Clone, Debug, Default, JsonSchema, EnumString, Display, PartialEq)]
#[serde(rename_all = "PascalCase")]
#[strum(serialize_all = "PascalCase")]
pub enum TriggerClusterSourceKind {
    #[default]
    ClusterGitRepo,
}

#[derive(Deserialize, Serialize, Clone, Debug, JsonSchema, PartialEq, Default)]
#[serde(default, rename_all = "camelCase")]
pub struct ActionJob {
    #[serde(skip_serializing_if = "Option::is_none")]
    workdir: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    cloner_image: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    action_image: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    args: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    command: Option<Vec<String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    service_account: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    annotations: Option<BTreeMap<String, String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    labels: Option<BTreeMap<String, String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    tolerations: Option<Vec<Toleration>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    affinity: Option<Affinity>,
    #[serde(skip_serializing_if = "Option::is_none")]
    node_selector: Option<BTreeMap<String, String>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    ttl_seconds_after_finished: Option<i32>,
    #[serde(skip_serializing_if = "Option::is_none")]
    active_deadline_seconds: Option<i64>,
    enable_cloner_debug: bool,
    preserve_git_folder: bool,
}

impl CustomApiResource for Action {
    fn kind(&self) -> &str {
        "Action"
    }
}

impl CustomApiResource for ClusterAction {
    fn kind(&self) -> &str {
        "ClusterAction"
    }
}

impl ActionExecutor for Action {}
impl ActionExecutor for ClusterAction {}

#[allow(private_bounds)]
pub(crate) trait ActionExecutor: ActionInternals {
    /// Creates K8s Job spec and run actual job from it
    #[allow(clippy::too_many_arguments)]
    async fn execute(
        &self,
        source_kind: &TriggerSourceKind,
        source_name: &str,
        source_commit: &str,
        trigger_ref: &TriggerGitRepoReference,
        client: Client,
        ns: &str,
        identity: String,
    ) -> Result<Job> {
        let job = self.create_job_spec(source_kind, source_name, source_commit, trigger_ref, ns, identity)?;

        info!("Create job {}/{}", ns, job.name_any());
        let jobs_api: Api<Job> = Api::namespaced(client.clone(), ns);
        jobs_api
            .create(&PostParams::default(), &job)
            .await
            .map_err(Error::KubeError)
    }
}

trait ActionInternals: Sized + Resource + CustomApiResource {
    fn action_job_spec(&self) -> ActionJob;
    fn source_override_spec(&self) -> Option<ActionSourceOverride>;
    fn get_owner_reference(&self) -> OwnerReference;

    /// format job name
    fn job_name(&self) -> String {
        let timestamp = DateTime::<Local>::from(SystemTime::now())
            .format("%Y%m%d-%H%M%S")
            .to_string();

        format!("{}-{}-{}", self.name_any(), timestamp, random_string(2)).to_lowercase()
    }

    /// just create K8s Job specification
    fn create_job_spec(
        &self,
        source_kind: &TriggerSourceKind,
        source_name: &str,
        source_commit: &str,
        trigger_ref: &TriggerGitRepoReference,
        ns: &str,
        identity: String,
    ) -> Result<Job> {
        let config = RuntimeConfig::get();
        let action_job = self.action_job_spec();
        let source_override = self.source_override_spec();

        // Create pod metadata if additional labels/annotations is defined
        let pod_template_metadata = if action_job.labels.is_some() || action_job.annotations.is_some() {
            let metadata = ObjectMeta {
                labels: action_job.labels.clone(),
                annotations: action_job.annotations.clone(),
                ..Default::default()
            };

            Some(metadata)
        } else {
            None
        };

        // Fill out custom jobs' labels
        let mut labels: BTreeMap<String, String> = BTreeMap::new();
        labels.insert(ACTION_JOB_IDENTITY_LABEL.into(), identity);
        labels.insert(ACTION_JOB_ACTION_KIND_LABEL.into(), self.kind().to_string());
        labels.insert(ACTION_JOB_ACTION_NAME_LABEL.into(), self.name_any());
        labels.insert(ACTION_JOB_SOURCE_KIND_LABEL.into(), source_kind.to_string());
        labels.insert(ACTION_JOB_SOURCE_NAME_LABEL.into(), source_name.into());

        // And add additional labels if defined
        if let Some(additional_labels) = action_job.labels {
            labels.extend(additional_labels);
        }

        let args = if let Some(source_override) = source_override {
            self.get_gitrepo_cloner_args(&source_override.kind, &source_override.name, &source_override.reference)
        } else {
            self.get_gitrepo_cloner_args(
                source_kind,
                source_name,
                &TriggerGitRepoReference::Commit(source_commit.into()),
            )
        };

        let service_account = if let Some(sa) = action_job.service_account {
            Some(sa)
        } else {
            RuntimeConfig::get().action.default_service_account.clone()
        };

        let ttl_seconds_after_finished = action_job
            .ttl_seconds_after_finished
            .unwrap_or(RuntimeConfig::get().action.ttl_seconds_after_finished);
        let active_deadline_seconds = action_job
            .active_deadline_seconds
            .unwrap_or(RuntimeConfig::get().action.active_deadline_seconds);

        let job = Job {
            metadata: ObjectMeta {
                name: Some(self.job_name()),
                namespace: Some(ns.to_string()),
                labels: Some(labels),
                annotations: action_job.annotations,
                owner_references: Some(vec![self.get_owner_reference()]),
                ..Default::default()
            },
            spec: Some(JobSpec {
                backoff_limit: Some(0),
                parallelism: Some(1),
                ttl_seconds_after_finished: Some(ttl_seconds_after_finished),
                active_deadline_seconds: Some(active_deadline_seconds),
                template: PodTemplateSpec {
                    metadata: pod_template_metadata,
                    spec: Some(PodSpec {
                        service_account_name: service_account,
                        node_selector: action_job.node_selector,
                        affinity: action_job.affinity,
                        tolerations: action_job.tolerations,
                        init_containers: Some(vec![Container {
                            name: config.action.containers.cloner.name.clone(),
                            image: Some(
                                action_job
                                    .cloner_image
                                    .unwrap_or(config.action.containers.cloner.image.clone()),
                            ),
                            args: Some(args),
                            volume_mounts: Some(vec![VolumeMount {
                                name: config.action.workdir.volume_name.clone(),
                                mount_path: action_job
                                    .workdir
                                    .clone()
                                    .unwrap_or(config.action.workdir.mount_path.clone()),
                                ..Default::default()
                            }]),
                            ..Default::default()
                        }]),
                        containers: vec![Container {
                            name: config.action.containers.worker.name.clone(),
                            image: Some(
                                action_job
                                    .action_image
                                    .unwrap_or(config.action.containers.worker.image.clone()),
                            ),
                            working_dir: Some(
                                action_job
                                    .workdir
                                    .clone()
                                    .unwrap_or(config.action.workdir.mount_path.clone()),
                            ),
                            command: action_job.command.clone(),
                            args: action_job.args,
                            env: Some(self.get_action_job_envs(
                                source_kind,
                                source_name,
                                source_commit,
                                trigger_ref,
                                ns,
                            )),
                            volume_mounts: Some(vec![VolumeMount {
                                name: config.action.workdir.volume_name.clone(),
                                mount_path: action_job.workdir.unwrap_or(config.action.workdir.mount_path.clone()),
                                ..Default::default()
                            }]),
                            ..Default::default()
                        }],
                        restart_policy: Some("Never".into()),
                        volumes: Some(vec![Volume {
                            name: config.action.workdir.volume_name.clone(),
                            empty_dir: Some(EmptyDirVolumeSource::default()),
                            ..Default::default()
                        }]),
                        ..Default::default()
                    }),
                },
                ..Default::default()
            }),
            ..Default::default()
        };

        Ok(job)
    }

    /// Prepare list of special environment variables to pass to worker container
    /// use `format_action_job_env_var` helper to unify approach
    fn get_action_job_envs(
        &self,
        source_kind: &TriggerSourceKind,
        source_name: &str,
        source_commit: &str,
        trigger_ref: &TriggerGitRepoReference,
        ns: &str,
    ) -> Vec<EnvVar> {
        let action_job = self.action_job_spec();
        let source_override = self.source_override_spec();

        let (trigger_ref_type, trigger_ref_name) = trigger_ref.to_refname(None);
        let mut envs = vec![
            format_action_job_env_var(
                "WORKDIR",
                &action_job
                    .workdir
                    .unwrap_or(RuntimeConfig::get().action.workdir.mount_path.clone()),
            ),
            format_action_job_env_var("TRIGGER_SOURCE_KIND", &source_kind.to_string()),
            format_action_job_env_var("TRIGGER_SOURCE_NAME", source_name),
            format_action_job_env_var("TRIGGER_SOURCE_COMMIT", source_commit),
            format_action_job_env_var("TRIGGER_SOURCE_REF_TYPE", &trigger_ref_type),
            format_action_job_env_var("TRIGGER_SOURCE_REF_NAME", trigger_ref_name),
        ];

        if source_kind == &TriggerSourceKind::GitRepo {
            envs.push(format_action_job_env_var("TRIGGER_SOURCE_NAMESPACE", ns));
        }

        if let Some(source_override) = source_override {
            let (ref_type, ref_name) = source_override.reference.to_refname(None);

            envs.push(format_action_job_env_var(
                "ACTION_SOURCE_KIND",
                &source_override.kind.to_string(),
            ));
            envs.push(format_action_job_env_var("ACTION_SOURCE_NAME", &source_override.name));
            envs.push(format_action_job_env_var("ACTION_SOURCE_REF_TYPE", ref_type.as_str()));
            envs.push(format_action_job_env_var("ACTION_SOURCE_REF_NAME", ref_name));
            if source_override.kind == TriggerSourceKind::GitRepo {
                envs.push(format_action_job_env_var("ACTION_SOURCE_NAMESPACE", ns));
            }
        }

        envs
    }

    /// prepares all needed arguments for cloner container entrypoint
    fn get_gitrepo_cloner_args(
        &self,
        source_kind: &TriggerSourceKind,
        source_name: &str,
        source_ref: &TriggerGitRepoReference,
    ) -> Vec<String> {
        let action_job = self.action_job_spec();

        let mut args = vec![
            "--kind".into(),
            source_kind.to_string(),
            "--source".into(),
            source_name.into(),
            "--destination".into(),
            action_job
                .workdir
                .clone()
                .unwrap_or(RuntimeConfig::get().action.workdir.mount_path.clone()),
        ];

        let (ref_type, ref_name) = source_ref.to_refname(Some("--"));
        args.push(ref_type);
        args.push(ref_name.clone());

        if self.action_job_spec().preserve_git_folder {
            args.push("--preserve-git-folder".into());
        }

        if action_job.enable_cloner_debug {
            args.push("--debug".into());
        }

        args
    }
}

impl ActionInternals for ClusterAction {
    fn action_job_spec(&self) -> ActionJob {
        self.spec.action_job.clone()
    }

    fn source_override_spec(&self) -> Option<ActionSourceOverride> {
        // Since ClusterAction restricts repo type to ClusterGitRepo only,
        // we create ActionSourceOverride from ClusterActionSourceOverride
        self.spec.source_override.as_ref().map(|value| ActionSourceOverride {
            kind: TriggerSourceKind::ClusterGitRepo,
            name: value.name.clone(),
            reference: value.reference.clone(),
        })
    }

    /// returns reference to action to set as owner to Job
    fn get_owner_reference(&self) -> OwnerReference {
        self.controller_owner_ref(&())
            .expect("unable to get owner reference, looks like a BUG!")
    }
}

impl ActionInternals for Action {
    fn action_job_spec(&self) -> ActionJob {
        self.spec.action_job.clone()
    }

    fn source_override_spec(&self) -> Option<ActionSourceOverride> {
        self.spec.source_override.clone()
    }

    fn get_owner_reference(&self) -> OwnerReference {
        self.controller_owner_ref(&())
            .expect("unable to get owner reference, looks like a BUG!")
    }
}

/// helper to create unified env var.
/// uses configured prefix to attach to each var.
fn format_action_job_env_var(name: &str, value: &str) -> EnvVar {
    EnvVar {
        name: format!(
            "{}{name}",
            RuntimeConfig::get().action.containers.worker.variables_prefix.clone()
        ),
        value: Some(value.to_owned()),
        value_from: None,
    }
}
