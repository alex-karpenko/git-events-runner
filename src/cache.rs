use crate::resources::{
    action::{Action, ClusterAction},
    git_repo::{ClusterGitRepo, GitRepo},
    trigger::{ScheduleTrigger, WebhookTrigger},
};
use crate::{Error, Result};
use core::hash::Hash;
use futures::future::ready;
use futures::StreamExt;
use k8s_openapi::{api::core::v1::Secret, ByteString};
use kube::runtime::WatchStreamExt;
use kube::{
    runtime::reflector::{self, Store},
    Api, Client,
};
use serde::Deserialize;
use std::{
    cmp::Eq,
    collections::{BTreeMap, HashMap},
    default::Default,
    fmt::Debug,
    marker::Send,
    sync::{Arc, OnceLock},
    time::{Duration, SystemTime},
};
use tokio::sync::RwLock;
use tracing::{debug, trace};

pub static SECRET_CACHE: OnceLock<SecretsCache> = OnceLock::new();

static API_CACHE_STORE: ApiCacheStore = ApiCacheStore {
    git_repo: OnceLock::new(),
    cluster_git_repo: OnceLock::new(),
    action: OnceLock::new(),
    cluster_action: OnceLock::new(),
    webhook_trigger: OnceLock::new(),
    schedule_trigger: OnceLock::new(),
};

pub struct ApiCacheStore {
    git_repo: OnceLock<Store<GitRepo>>,
    cluster_git_repo: OnceLock<Store<ClusterGitRepo>>,
    action: OnceLock<Store<Action>>,
    cluster_action: OnceLock<Store<ClusterAction>>,
    webhook_trigger: OnceLock<Store<WebhookTrigger>>,
    schedule_trigger: OnceLock<Store<ScheduleTrigger>>,
}

impl ApiCacheStore {
    pub async fn watch(client: Client) {
        tokio::join!(
            GitRepo::watch(client.clone()),
            ClusterGitRepo::watch(client.clone()),
            Action::watch(client.clone()),
            ClusterAction::watch(client.clone()),
            WebhookTrigger::watch(client.clone()),
            ScheduleTrigger::watch(client.clone()),
        );
    }
}

#[allow(async_fn_in_trait)]
pub trait ApiCache {
    fn get(name: &str, namespace: Option<&str>) -> Result<Arc<Self>>
    where
        Self: Sized,
        Self: 'static + reflector::Lookup,
        Self::DynamicType: Hash + Eq,
        Self: Clone,
        <Self as reflector::Lookup>::DynamicType: Clone,
        <Self as reflector::Lookup>::DynamicType: Default,
    {
        let store = Self::get_cache_store();
        let key = reflector::ObjectRef::new(name).within(namespace.unwrap_or(&String::from("default")));
        if let Some(value) = store.get(&key) {
            Ok(value)
        } else {
            let name = if let Some(namespace) = namespace {
                format!("{namespace}/{name}")
            } else {
                name.to_string()
            };
            Err(Error::ResourceNotFoundError(name))
        }
    }

    async fn watch(client: Client)
    where
        Self: 'static + Sized + Clone + Debug + Send,
        Self: kube::Resource,
        <Self as kube::Resource>::DynamicType: Default + Eq + Hash + Clone,
        for<'de> Self: Deserialize<'de>,
    {
        let api: Api<Self> = Api::all(client);
        let (store, writer) = reflector::store();
        let stream = reflector::reflector(
            writer,
            kube::runtime::watcher(api, kube::runtime::watcher::Config::default()),
        );

        Self::set_cache_store(store);
        let watch = stream.applied_objects().for_each(|o| {
            trace!(object = ?o, "updating object in cache");
            ready(())
        });
        watch.await
    }

    fn get_cache_store() -> Store<Self>
    where
        Self: Sized,
        Self: 'static + reflector::Lookup,
        Self::DynamicType: Hash + Eq;

    fn set_cache_store(store: Store<Self>)
    where
        Self: Sized,
        Self: 'static + reflector::Lookup,
        Self::DynamicType: Hash + Eq;
}

impl ApiCache for GitRepo {
    fn get_cache_store() -> Store<Self> {
        API_CACHE_STORE.git_repo.get().unwrap().clone()
    }

    fn set_cache_store(store: Store<Self>) {
        API_CACHE_STORE.git_repo.set(store).unwrap();
    }

    #[cfg(test)]
    fn get(name: &str, namespace: Option<&str>) -> Result<Arc<Self>>
    where
        Self: Sized + Clone,
    {
        use std::thread::spawn;
        use tokio::runtime::Runtime;

        spawn({
            let name = String::from(name);
            let namespace = namespace.map(String::from);

            move || {
                let rt = Runtime::new()?;
                rt.block_on(async move {
                    let client = Client::try_default().await.unwrap();
                    let api = Api::<Self>::namespaced(client, &namespace.unwrap());
                    match api.get(&name).await {
                        Ok(resource) => Ok(Arc::new(resource)),
                        Err(e) => Err(e.into()),
                    }
                })
            }
        })
        .join()
        .unwrap()
    }
}

impl ApiCache for ClusterGitRepo {
    fn get_cache_store() -> Store<Self> {
        API_CACHE_STORE.cluster_git_repo.get().unwrap().clone()
    }

    fn set_cache_store(store: Store<Self>) {
        API_CACHE_STORE.cluster_git_repo.set(store).unwrap();
    }
}

impl ApiCache for Action {
    fn get_cache_store() -> Store<Self> {
        API_CACHE_STORE.action.get().unwrap().clone()
    }

    fn set_cache_store(store: Store<Self>) {
        API_CACHE_STORE.action.set(store).unwrap();
    }
}

impl ApiCache for ClusterAction {
    fn get_cache_store() -> Store<Self> {
        API_CACHE_STORE.cluster_action.get().unwrap().clone()
    }

    fn set_cache_store(store: Store<Self>) {
        API_CACHE_STORE.cluster_action.set(store).unwrap();
    }
}

impl ApiCache for WebhookTrigger {
    fn get_cache_store() -> Store<Self> {
        API_CACHE_STORE.webhook_trigger.get().unwrap().clone()
    }

    fn set_cache_store(store: Store<Self>) {
        API_CACHE_STORE.webhook_trigger.set(store).unwrap();
    }

    #[cfg(test)]
    fn get(name: &str, namespace: Option<&str>) -> Result<Arc<Self>>
    where
        Self: Sized + Clone,
    {
        use std::thread::spawn;
        use tokio::runtime::Runtime;

        spawn({
            let name = String::from(name);
            let namespace = namespace.map(String::from);

            move || {
                let rt = Runtime::new()?;
                rt.block_on(async move {
                    let client = Client::try_default().await.unwrap();
                    let api = Api::<Self>::namespaced(client, &namespace.unwrap());
                    match api.get(&name).await {
                        Ok(resource) => Ok(Arc::new(resource)),
                        Err(e) => Err(e.into()),
                    }
                })
            }
        })
        .join()
        .unwrap()
    }
}

impl ApiCache for ScheduleTrigger {
    fn get_cache_store() -> Store<Self> {
        API_CACHE_STORE.schedule_trigger.get().unwrap().clone()
    }

    fn set_cache_store(store: Store<Self>) {
        API_CACHE_STORE.schedule_trigger.set(store).unwrap();
    }

    #[cfg(test)]
    fn get(name: &str, namespace: Option<&str>) -> Result<Arc<Self>>
    where
        Self: Sized + Clone,
    {
        use std::thread::spawn;
        use tokio::runtime::Runtime;

        spawn({
            let name = String::from(name);
            let namespace = namespace.map(String::from);

            move || {
                let rt = Runtime::new()?;
                rt.block_on(async move {
                    let client = Client::try_default().await.unwrap();
                    let api = Api::<Self>::namespaced(client, &namespace.unwrap());
                    match api.get(&name).await {
                        Ok(resource) => Ok(Arc::new(resource)),
                        Err(e) => Err(e.into()),
                    }
                })
            }
        })
        .join()
        .unwrap()
    }
}

#[derive(Clone)]
pub struct SecretsCache {
    expiration_timeout: Duration,
    client: Client,
    cache: Arc<RwLock<HashMap<String, SecretValue>>>,
}

#[derive(Debug)]
struct SecretValue {
    data: BTreeMap<String, ByteString>,
    expires_at: SystemTime,
}

impl Debug for SecretsCache {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("ExpiringSecretCache")
            .field("expiration_timeout", &self.expiration_timeout)
            .field("cache", &self.cache)
            .finish()
    }
}

impl SecretsCache {
    #[cfg(not(test))]
    pub async fn get(namespace: &str, secret_name: &str, key: &str) -> Result<String, Error> {
        SECRET_CACHE
            .get()
            .expect("unable to get shared secret cache instance, looks like a BUG!")
            .get_key_value(namespace, secret_name, key)
            .await
    }

    #[cfg(test)]
    pub async fn get(namespace: &str, secret_name: &str, key: &str) -> Result<String, Error> {
        let client = Client::try_default().await.unwrap();
        let (_, secret_value) = Self::query_secrets_api(client, namespace, secret_name, key).await?;

        Ok(secret_value)
    }

    /// Creates shared static instance of the cache
    pub fn init_cache(expiration_timeout: Duration, client: Client) {
        SECRET_CACHE
            .set(Self {
                expiration_timeout,
                client,
                cache: Arc::new(RwLock::new(HashMap::new())),
            })
            .expect("failed to init shared secret cache, looks like a BUG!");
    }

    /// Does actual work with secrets and hides implementation
    async fn get_key_value(&self, namespace: &str, secret_name: &str, key: &str) -> Result<String, Error> {
        let hash_key = format!("{namespace}/{secret_name}");
        debug!(secret = %hash_key, %key, "getting secret from cache");

        // Let's try to find in cache
        {
            let cache = self.cache.read().await;
            if let Some(secret) = cache.get(&hash_key) {
                // and return value if it's cached and not expired and contains key
                let value = secret.data.get(key);
                if secret.expires_at > SystemTime::now() && value.is_some() {
                    let value = value.unwrap().0.clone();
                    let value = String::from_utf8(value).map_err(|_e| {
                        Error::SecretDecodingError(format!(
                            "error converting string `{key}` from UTF8 in the secret `{secret_name}`"
                        ))
                    })?;
                    trace!(secret = %hash_key, %key, "object found");
                    return Ok(value);
                } else {
                    trace!(secret = %hash_key, %key, "object expired or key doesn't exist");
                }
            } else {
                trace!(secret = %hash_key, %key, "object is not in the cache");
            }
        }

        // If it's not cached yet or already expired - retrieve secret from API and store to cache
        let mut cache = self.cache.write().await;
        trace!(secret = %hash_key, %key, "try to retrieve and save to the cache");
        let (secret_data_raw, secret_value) =
            Self::query_secrets_api(self.client.clone(), namespace, secret_name, key).await?;

        let cache_data = SecretValue {
            data: secret_data_raw,
            expires_at: SystemTime::now()
                .checked_add(self.expiration_timeout)
                .expect("looks like a BUG!"),
        };
        trace!(secret = %hash_key, %key, "save object to the cache");
        cache.insert(hash_key, cache_data);

        Ok(secret_value)
    }

    pub async fn query_secrets_api(
        client: Client,
        namespace: &str,
        secret_name: &str,
        key: &str,
    ) -> Result<(BTreeMap<String, ByteString>, String), Error> {
        let secrets_api: Api<Secret> = Api::namespaced(client, namespace);
        let secret = secrets_api.get(secret_name).await?;

        let secret_data_raw = secret
            .clone()
            .data
            .ok_or_else(|| Error::SecretDecodingError(format!("no `data` part in the secret `{secret_name}`")))?;
        let secret_data = secret_data_raw
            .get(key)
            .ok_or_else(|| Error::SecretDecodingError(format!("no `{key}` key in the secret `{secret_name}`")))?
            .to_owned();
        let secret_value = String::from_utf8(secret_data.0).map_err(|_e| {
            Error::SecretDecodingError(format!(
                "error converting string `{key}` from UTF8 in the secret `{secret_name}`"
            ))
        })?;

        Ok((secret_data_raw, secret_value))
    }
}
