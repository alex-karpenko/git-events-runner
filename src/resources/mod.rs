//! Custom resources definitions
pub mod action;
pub mod git_repo;
pub mod trigger;

use crate::{controller::Context, Result};
use action::{Action, ClusterAction};
use git_repo::{ClusterGitRepo, GitRepo};
use k8s_openapi::apiextensions_apiserver::pkg::apis::apiextensions::v1::CustomResourceDefinition;
use kube::{runtime::controller::Action as ReconcileAction, CustomResourceExt};
use rand::{distributions::Alphanumeric, thread_rng, Rng};
use std::sync::Arc;
use trigger::{ScheduleTrigger, WebhookTrigger};

const API_GROUP: &str = "git-events-runner.rs";
const CURRENT_API_VERSION: &str = "v1alpha1";

pub(crate) trait CustomApiResource {
    fn crd_kind() -> &'static str;
}

/// Behavior of all CRDs that have external state and should be able to reconcile
#[allow(async_fn_in_trait)]
pub trait Reconcilable<S> {
    /// Reconciles `self`
    async fn reconcile(&self, ctx: Arc<Context>) -> Result<ReconcileAction>;
    /// Finalizes `self`
    async fn cleanup(&self, ctx: Arc<Context>) -> Result<ReconcileAction>;
    /// Returns finalizer name
    fn finalizer_name(&self) -> Option<&'static str> {
        None
    }
}

pub(crate) fn random_string(len: usize) -> String {
    let rand: String = thread_rng()
        .sample_iter(&Alphanumeric)
        .take(len)
        .map(char::from)
        .collect();
    rand
}

/// Return list with all CRD definitions
pub fn get_all_crds() -> Vec<CustomResourceDefinition> {
    vec![
        GitRepo::crd(),
        ClusterGitRepo::crd(),
        ScheduleTrigger::crd(),
        WebhookTrigger::crd(),
        Action::crd(),
        ClusterAction::crd(),
    ]
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::HashSet;

    #[test]
    fn ensure_100_4ch_random_strings() {
        const SET_SIZE: usize = 100;

        let mut control_set = HashSet::new();
        for _ in 0..SET_SIZE {
            control_set.insert(random_string(4));
        }

        assert_eq!(control_set.len(), SET_SIZE);
    }
}
