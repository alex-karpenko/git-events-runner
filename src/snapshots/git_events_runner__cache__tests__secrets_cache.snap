---
source: src/cache.rs
expression: s.err().unwrap()
---
KubeError(
    Api(
        ErrorResponse {
            status: "Failure",
            message: "secrets \"test-secrets-cache-data\" not found",
            reason: "NotFound",
            code: 404,
        },
    ),
)
